### PR TITLE
[SPARK-28903][STREAMING][PYSPARK][TESTS] Fix AWS JDK version conflict that breaks Pyspark Kinesis tests

### PR DESCRIPTION
**Pull request by srowen**
The Pyspark Kinesis tests are failing, at least in master:

```
======================================================================
ERROR: test_kinesis_stream (pyspark.streaming.tests.test_kinesis.KinesisStreamTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/pyspark/streaming/tests/test_kinesis.py", line 44, in test_kinesis_stream
    kinesisTestUtils = self.ssc._jvm.org.apache.spark.streaming.kinesis.KinesisTestUtils(2)
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/lib/py4j-0.10.8.1-src.zip/py4j/java_gateway.py", line 1554, in __call__
    answer, self._gateway_client, None, self._fqn)
  File "/home/jenkins/workspace/SparkPullRequestBuilder@2/python/lib/py4j-0.10.8.1-src.zip/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
Py4JJavaError: An error occurred while calling None.org.apache.spark.streaming.kinesis.KinesisTestUtils.
: java.lang.NoSuchMethodError: com.amazonaws.regions.Region.getAvailableEndpoints()Ljava/util/Collection;
	at org.apache.spark.streaming.kinesis.KinesisTestUtils$.$anonfun$getRegionNameByEndpoint$1(KinesisTestUtils.scala:211)
	at org.apache.spark.streaming.kinesis.KinesisTestUtils$.$anonfun$getRegionNameByEndpoint$1$adapted(KinesisTestUtils.scala:211)
	at scala.collection.Iterator.find(Iterator.scala:993)
	at scala.collection.Iterator.find$(Iterator.scala:990)
	at scala.collection.AbstractIterator.find(Iterator.scala:1429)
	at scala.collection.IterableLike.find(IterableLike.scala:81)
	at scala.collection.IterableLike.find$(IterableLike.scala:80)
	at scala.collection.AbstractIterable.find(Iterable.scala:56)
	at org.apache.spark.streaming.kinesis.KinesisTestUtils$.getRegionNameByEndpoint(KinesisTestUtils.scala:211)
	at org.apache.spark.streaming.kinesis.KinesisTestUtils.<init>(KinesisTestUtils.scala:46)
...
```

The non-Python Kinesis tests are fine though. It turns out that this is because Pyspark tests use the output of the Spark assembly, and it pulls in `hadoop-cloud`, which in turn pulls in an old AWS Java SDK.

Per Steve Loughran (below), it seems like we can just resolve this by excluding the aws-java-sdk dependency. See the attached PR for some more detail about the debugging and other options.

See https://github.com/apache/spark/pull/25558#issuecomment-524042709

### CODE REVIEW

**Comment by srowen** [comment]
I'm pretty stumped on this. The NoSuchMethodError refers to a method that is definitely there. The AWS dependencies are harmonized. The Scala tests work. The dependency graph from SBT looks right. I am guessing something about how python tests work or are run is the issue, but not sure what would cause this, as it suggests a runtime AWS SDK version difference, but I don't see any other versions pulled in.

**Comment by sarutak** [comment]
The reason for failure of the unit test in #24651 seems to be same as you hit...

File: hadoop-cloud/pom.xml [hunk 1]
```diff
@@ -100,6 +100,10 @@
           <groupId>com.fasterxml.jackson.core</groupId>
           <artifactId>jackson-annotations</artifactId>
         </exclusion>
+        <exclusion>
```
[the remark this answers is not in the document]
Comment (steveloughran) [reply hadoop-cloud/pom.xml]: 1.7.4 is a really old version; 2.9+ uses a (fat) shaded JAR which has a consistent kinesis SDK in with it

[the remark this answers is not in the document]
Comment (steveloughran) [reply hadoop-cloud/pom.xml]: +1 to excluding the AWS dependency. It is not actually something you can bundle into ASF releases anyway [https://issues.apache.org/jira/browse/HADOOP-13794](https://issues.apache.org/jira/browse/HADOOP-13794). But: it'd be good for a spark-hadoop-cloud artifact to be published with that dependency for downstream users, or at least the things you have to add documented somewhere.

FWIW,  I do build and test the spark kinesis module as part of my AWS SDK update process -one that actually went pretty smoothly for a change last time. No regressions, no new error messages in logs, shaded JARs really are shaded, etc. This is progress and means that backporting is something we should be doing

see https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/testing.md#-qualifying-an-aws-sdk-update for the runbook there

**Review by srowen — COMMENTED** [review]
@steveloughran I'll re-run tests without my placeholder TODOs, but it won't matter (Kinesis tests won't run anyway). Looks like it passes with the exclusion. That's OK by you?

**Comment by steveloughran** [comment]
could some stick up the output of the mvn dependency listing here `mvn dependency:tree -Dverbose` for the base profile and the hadoop-3.1 branch? I'll be able to review them.

**Comment by srowen** [comment]
@steveloughran does that seem reasonable? the above is after removing `aws-java-sdk` per your suggestion. That definitely fixes the issue and would be good to get these tests back online. However I don't want to fundamentally break the purpose of `hadoop-cloud`. Would a user of this be running on Hadoop and already have this stuff on the classpath?

**Comment by srowen** [comment]
Per https://github.com/apache/spark/pull/25559#discussion_r318557499 I'm going to proceed, to make the tests work again. If we need to make a more nuanced change we can do so in a follow-up.

**Comment by srowen** [comment]
Merged to master/2.4. If this turns out to cause problems for `hadoop-cloud`'s usage, we can look at another solution like instead including this (and all its transitive dependencies) as a direct dependency, and harmonize the version. I wasn't clear whether that was necessary given how it's used, nor if it's even OK to push the SDK version much higher for 2.7. Sounds like it's kind of a non-issue in 2.9+.

**Comment by dongjoon-hyun** [comment]
Thank you for this fix, @srowen !

**Comment by steveloughran** [comment]
sorry, been in hiding. hadoop branch-2 and cloud stuff is trouble as the aws sdk support has had to chase a moving target dependency-wise.

* if the hadoop-cloud-storage pom/jar isn't consistent in branch-2, then, well, so be it. It's time to move off hadoop 2.x anyway.
* I do want it to work on hadoop 3.2; if that's no longer working then that can be fixed. It does include the full shaded AWS SDK so it's unlikely that you'd get inconsistencies across modules; it's also made upgrading that SDK easier, especially as the SDK classes haven't made any incompatible changes for a while (mostly)

FWIW we use the apache hadoop-cloud-storage POM as the source of truth for which cloud store bits dependent apps pick up (spark, hive, etc); the spark hadoop-cloud POM pulls that in and then gets it into the spark releases. This provides a one-stop pipeline to get in things which aren't normally in releases (Google GCS) and leave out bits which aren't currently supports (allyun OSS). That generally keeps things under control, leaving only configuration settings...

**Pull request closed without merging**